### PR TITLE
[enhancement] add optional (features) fields in daemon.json to enable buildkit

### PR DESCRIPTION
**Pull request by AntaresS**
Signed-off-by: Anda Xu <anda.xu@docker.com>

**- What I did**
- Add optional fields to allow enable buildkit as Dockerfile builder from daemon.json.
  - Add new flag `--buildkit` in `dockerd` to enable buildkit as Dockerfile builder. (This is also a pre-req by the daemon implementaion)
- no longer assuming buildkit is on by default if "--experimental" is true.

**- How I did it**

**- How to verify it**
image

also `dockerd --help` will print help info of `--buildkit` flag
```
 --buildkit                                Enable buildkit as dockerfile builder
```
**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by AntaresS** [comment]
@tiborvass @dhiltgen ptal

**Comment by AntaresS** [comment]
cc @andrewhsu

**Comment by tonistiigi** [comment]
@tiborvass I don't think we need a flag. Config should be enough.

File: cmd/dockerd/daemon.go [hunk 1]
```diff
@@ -494,6 +494,14 @@ func initRouter(opts routerOptions) {
 func (cli *DaemonCli) initMiddlewares(s *apiserver.Server, cfg *apiserver.Config, pluginStore plugingetter.PluginGetter) error {
 	v := cfg.Version
 
+	buildkitEnabled := "false"
+	if cli.Config.Buildkit {
+		buildkitEnabled = "true"
+	}
+	specialHeaders := middleware.SpecialHeaders{middleware.BuildkitHeader: buildkitEnabled}
+	special := middleware.NewSpecialMiddleware(specialHeaders)
```
Comment (tonistiigi) [inline cmd/dockerd/daemon.go]: Why is this a middleware?

Comment (tiborvass) [reply cmd/dockerd/daemon.go]: I guess it's because all the other things that were setting headers were middlewares as well.

Comment (AntaresS) [reply cmd/dockerd/daemon.go]: @tonistiigi This is how the request headers are set, `Docker-Experimental` for one example. In an early discussion, we wanted to have `_ping` endpoint to carry the information of whether buildkit is enabled on daemon. And this middleware sets the header to `_ping`.

**Comment by AntaresS** [comment]
> @tiborvass I don't think we need a flag. Config should be enough.

@tonistiigi I thought this way too. However, it seems that when the daemon starts and loads the config, there is always a flag check to make sure whatever in the `daemon.json` is also a valid flag. Otherwise, an error will be returned by https://github.com/moby/moby/blob/f57f260b49b6142366e6bc1274204ee0a1205945/daemon/config/config.go#L470

File: api/types/types.go [hunk 2]
```diff
@@ -105,6 +105,7 @@ type Ping struct {
 	APIVersion   string
 	OSType       string
 	Experimental bool
+	Buildkit     bool
```
Comment (tiborvass) [inline api/types/types.go]: `BuildKit` that's how it's spelled elsewhere especially for API types

Comment (tonistiigi) [reply api/types/types.go]: Should it be `FeatureBuildKit` ? Also, need to make the flag longer, or modify the config condition.

File: cmd/dockerd/daemon.go (hunk 1 above)
Comment (tonistiigi) [reply cmd/dockerd/daemon.go]: This looks very weird that API command that does not have anything to do with builder returns `Buildkit=true`.

**Comment by AntaresS** [comment]
@tonistiigi @tiborvass Made "buildkit" an exception from flag validation. Now it's only an optional field in daemon.json

File: api/server/middleware/special.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	"net/http"
+)
+
+const (
+	// BuildKitHeader is the key of buildkit header
+	BuildKitHeader = "BuildKit-Enabled"
+)
+
+// SpecialHeaders contains the key value pairs to be set as headers
+type SpecialHeaders map[string]string
```
Comment (cpuguy83) [inline api/server/middleware/special.go]: What makes these special as opposed to just headers?

File: api/server/middleware/special.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package middleware
+
+import (
+	"context"
+	"net/http"
+)
+
+const (
+	// BuildKitHeader is the key of buildkit header
+	BuildKitHeader = "BuildKit-Enabled"
```
Comment (cpuguy83) [inline api/server/middleware/special.go]: Worried that we are leaking implementation details into the API package here.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
I don't want to over-complicate this, but it seems we are now moving from a generic all-or-nothing flag to feature flags (liberal use of the term `flag` here).

I would expect that the pieces in the API to signal somehow that something is a feature flag rather than just some top-level field.

File: api/types/types.go [hunk 5]
```diff
@@ -105,6 +105,7 @@ type Ping struct {
 	APIVersion   string
 	OSType       string
 	Experimental bool
+	BuildKit     bool
```
Comment (tiborvass) [inline api/types/types.go]: `FeatureBuildKit`

File: cmd/dockerd/daemon.go [hunk 6]
```diff
@@ -494,6 +494,14 @@ func initRouter(opts routerOptions) {
 func (cli *DaemonCli) initMiddlewares(s *apiserver.Server, cfg *apiserver.Config, pluginStore plugingetter.PluginGetter) error {
 	v := cfg.Version
 
+	buildkitEnabled := "false"
+	if cli.Config.BuildKit != nil && *cli.Config.BuildKit {
+		buildkitEnabled = "true"
+	}
+	specialHeaders := middleware.SpecialHeaders{middleware.BuildKitHeader: buildkitEnabled}
+	special := middleware.NewSpecialMiddleware(specialHeaders)
+	s.UseMiddleware(special)
```
[the remark this answers is not in the document]
Comment (AntaresS) [reply cmd/dockerd/daemon.go]: @tiborvass well the ping handler by itself doesn't know anything about daemon config. Unless we modify the `Backend` interface to get the info of whether buildkit is enabled or somehow to pass that knowledge into `NewRouter`. Is this more desired?

File: api/types/types.go (hunk 5 above)
Comment (AntaresS) [reply api/types/types.go]: Done

**Comment by AntaresS** [comment]
@tiborvass @tonistiigi ptal

**Comment by AntaresS** [comment]
```

**Comment by andrewhsu** [comment]
@AntaresS the test-docker-py failures from [janky](https://jenkins.dockerproject.org/job/Docker-PRs/50365/console) run of a0d285f are reproducible in my dev env:
```
$ make test-docker-py
...
=== 63 failed, 250 passed, 5 skipped, 5 xfailed, 15 error in 298.31 seconds ====
```

seems to hang on that last line, need to control-c out of it

File: api/server/router/system/system_routes.go [hunk 7]
```diff
@@ -25,7 +25,11 @@ func optionsHandler(ctx context.Context, w http.ResponseWriter, r *http.Request,
 	return nil
 }
 
-func pingHandler(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+func (s *systemRouter) pingHandler(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+	w.Header().Set("BuildKit-Enabled", "false")
```
Comment (tiborvass) [inline api/server/router/system/system_routes.go]: `Feature-BuildKit-Enabled`

File: cmd/dockerd/daemon.go [hunk 8]
```diff
@@ -271,17 +271,19 @@ func newRouterOptions(config *config.Config, daemon *daemon.Daemon) (routerOptio
...
 	}
-
-	buildkit, err := buildkit.New(buildkit.Opt{
-		SessionManager: sm,
-		Root:           filepath.Join(config.Root, "buildkit"),
-		Dist:           daemon.DistributionServices(),
-	})
-	if err != nil {
-		return opts, err
+	var builder *buildkit.Builder
```
Comment (tiborvass) [inline cmd/dockerd/daemon.go]: you can call the variable `bk` since this is specifically the buildkit builder. Otherwise it's weird to see `builder` being `nil` when buildkit is disabled.

File: client/ping.go [hunk 9]
```diff
@@ -7,7 +7,7 @@ import (
 	"github.com/docker/docker/api/types"
 )
 
-// Ping pings the server and returns the value of the "Docker-Experimental", "OS-Type" & "API-Version" headers
+// Ping pings the server and returns the value of the "Docker-Experimental", "BuildKit-Enabled", "OS-Type" & "API-Version" headers
```
Comment (tiborvass) [inline client/ping.go]: `Feature-BuildKit-Enabled`

**Review by tiborvass — CHANGES_REQUESTED** [review]

**Comment by tiborvass** [comment]
@cpuguy83 
> I would expect that the pieces in the API to signal somehow that something is a feature flag rather than just some top-level field.

Were you mentioning the API in the broader sense (e.g., the config-file setting)?
IOW, I'm undecided on the `-Enabled` suffix in the `Feature-BuildKit-Enabled` HTTP header, given that it'd be inconsistent with the `feature-buildkit` config setting which doesn't have the suffix, and that `feature-buildkit-enabled` is long.

So either:
1. It's fine as-is
2. we remove `-enabled` from HTTP header
3. we add `-enabled` to config setting

**Comment by AntaresS** [comment]
All tests passed. Now rebasing.
cc @andrewhsu

**Comment by AntaresS** [comment]
@tiborvass Just amended the PR. The original tristate implementation is not thorough, which only sets the `FeatureBuildkit` state at reading in config level but still using it as a binary state in all the other places. This also explains the tests failures.

File: api/server/router/system/system_routes.go [hunk 10]
```diff
@@ -25,7 +25,12 @@ func optionsHandler(ctx context.Context, w http.ResponseWriter, r *http.Request,
 	return nil
 }
 
-func pingHandler(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+func (s *systemRouter) pingHandler(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+	if s.featureBuildKit != nil && *s.featureBuildKit {
```
Comment (tiborvass) [inline api/server/router/system/system_routes.go]:
```Go
if s.featureBuildKit != nil {
    s := "false"
    if *s.featureBuildKit {
        s = "true"
    }
    w.Header().Set("Feature-BuildKit-Enabled", s)
}
```

File: client/ping.go [hunk 11]
```diff
@@ -27,6 +27,13 @@ func (cli *Client) Ping(ctx context.Context) (types.Ping, error) {
 			ping.Experimental = true
 		}
 		ping.OSType = serverResp.header.Get("OSType")
+		if serverResp.header.Get("Feature-BuildKit-Enabled") == "true" {
```
Comment (tiborvass) [inline client/ping.go]: Use switch

File: api/server/router/system/system_routes.go (hunk 10 above)
Comment (vdemeester) [reply api/server/router/system/system_routes.go]: This can be simplified even further : 

```go
if s.featureBuildkit != nil {
    w.Header().Set("Feature-BuildKit-Enabled", fmt.Sprintf("%v", *s.featureBuildkit))
}
```

File: api/types/types.go [hunk 12]
```diff
@@ -102,9 +102,10 @@ type ContainerStats struct {
 // Ping contains response of Engine API:
 // GET "/_ping"
 type Ping struct {
-	APIVersion   string
-	OSType       string
-	Experimental bool
+	APIVersion      string
+	OSType          string
+	Experimental    bool
+	FeatureBuildKit *bool
```
Comment (vdemeester) [inline api/types/types.go]: If we want *more* feature flag in the future, maybe we should make this one a struct or a map 👼 

```
Features map[string]bool
```

**Review by vdemeester — CHANGES_REQUESTED** [review]
Few comments 👼

File: api/types/types.go (hunk 12 above)
Comment (cpuguy83) [reply api/types/types.go]: *cough* `map[string]string` *cough*

Comment (AntaresS) [reply api/types/types.go]: @cpuguy83 I think `map[string]bool` is actually better here, which will satisfy the tri-state we discussed earlier.

Comment (cpuguy83) [reply api/types/types.go]: The reason I say string is because you can change strings and represent quite a bit more.... but sounds like people think it may be overkill. That's fine.

File: api/server/router/system/system_routes.go (hunk 10 above)
Comment (AntaresS) [reply api/server/router/system/system_routes.go]: @tiborvass done

File: client/ping.go (hunk 11 above)
Comment (AntaresS) [reply client/ping.go]: done

**Comment by AntaresS** [comment]
Changes have been made accordingly, lgty? @tiborvass @vdemeester @cpuguy83 
cc @andrewhsu

File: api/types/types.go [hunk 13]
```diff
@@ -102,9 +102,10 @@ type ContainerStats struct {
 // Ping contains response of Engine API:
 // GET "/_ping"
 type Ping struct {
-	APIVersion   string
-	OSType       string
-	Experimental bool
+	APIVersion      string
+	OSType          string
+	Experimental    bool
+	FeatureBuildKit map[string]bool
```
[the remark this answers is not in the document]
Comment (AntaresS) [reply api/types/types.go]: yes it should

**Comment by AntaresS** [comment]
@tiborvass Throw some thoughts on the first bullet point. The experimental header from the HTTP response is actually `Docker-Experimental` (showing in the screenshot above), and IMHO to make it consistent I think `Feature-BuildKit` is simpler and self-explanatory. Also consider that we don't have `Docker-Experimental-Enabled`.

**Comment by vdemeester** [comment]
> 1. Config is named `feature-buildkit` vs header is named `Feature-BuildKit-Enabled`. My understanding is that the `-Enabled` suffix comes from being consistent with the `Experimental-Enabled` header. Is this fine?

I think `Feature-Buildkit` header is fine (and would allow to pass either some binary option `enable`/`disable` or more)

> 2. We're detecting whether buildkit is enabled, but in the `/build` endpoint we use a version to distinguish between legacy (v1) and buildkit (v2). The discussion about `map[string]string` reminded me of this. Do you feel this is awkward or it's fine?

I think it's fine

> 3. If we decide on `map[string]bool` for the struct field, shouldn't it be called `Features`? In which case, shouldn't the config also be called `features`? The user would then have to write `"features": {"buildkit": true}`.

I think `map[string]string` would be even better (sorry for the bikeshed a bit @AntaresS :pray:). And yes, *client-side*, I think we should use `features` too and slowly deprecated the `experimental` one (i.e. moving from a all-or-nothing experimental to feature toggles).

@tiborvass ^^

**Comment by tiborvass** [comment]
@cpuguy83 @vdemeester @dmcgowan 

Let's keep the API somewhat consistent with itself, so let's not use "feature knobs" but instead pass in the default builder version. We already have a `version` param on `/build`: `1` is old builder, and `2` is buildkit.

Really sorry for all the back and forth, naming is hard, this is my final design review: if the daemon has buildkit enabled, it sets `Builder-Default-Version` header in the ping response, suggesting that it can be overridden. Otherwise, the header is not sent, and if buildkit was disabled but a client tries using buildkit on `/build` it will receive an error from the daemon. If nothing was specified on the daemon, then client can honor `DOCKER_BUILDKIT` envvar without errors.

For the daemon config, I'm fine with `"feature-buildkit": true`/`"feature-buildkit": false`. It's not about specifying the builder to use, nor always about the default builder to use, it's a mixture between default builder and disallowing the use of buildkit. Also from a UX perspective it's easy to think of it as just enabling/disabling the buildkit feature daemon-wide.

Please let me know quickly if you disagree. Otherwise I think we'll just go with this.

**Comment by cpuguy83** [comment]
SGTM

**Comment by vdemeester** [comment]
SGTM 😉

File: api/server/router/system/system.go [hunk 14]
```diff
@@ -9,25 +9,27 @@ import (
...
+	cluster         ClusterBackend
+	routes          []router.Route
+	fscache         *fscache.FSCache // legacy
+	builder         *buildkit.Builder
+	featureBuildKit *bool
 }
 
 // NewRouter initializes a new system router
-func NewRouter(b Backend, c ClusterBackend, fscache *fscache.FSCache, builder *buildkit.Builder) router.Router {
+func NewRouter(b Backend, c ClusterBackend, fscache *fscache.FSCache, builder *buildkit.Builder, featureBuildKit *bool) router.Router {
```
Comment (thaJeztah) [inline api/server/router/system/system.go]: Not a fan of the `featureBuildKit` boolean; can't we make this a `features []string` or `features map[string]string` ?

File: api/types/types.go [hunk 15]
```diff
@@ -105,6 +105,7 @@ type Ping struct {
 	APIVersion   string
 	OSType       string
 	Experimental bool
+	Features     map[string]bool
```
Comment (thaJeztah) [inline api/types/types.go]: Unless we want to use this to _disable_ features that are enabled by default, we could simply use a `[]string` (i.e. presence of a feature just means it's in the list?)

File: cmd/dockerd/config.go [hunk 16]
```diff
@@ -65,7 +65,6 @@ func installCommonConfigFlags(conf *config.Config, flags *pflag.FlagSet) {
 
 	flags.StringVar(&conf.SwarmDefaultAdvertiseAddr, "swarm-default-advertise-addr", "", "Set default address or interface for swarm advertised address")
 	flags.BoolVar(&conf.Experimental, "experimental", false, "Enable experimental features")
-
```
Comment (thaJeztah) [inline cmd/dockerd/config.go]: unrelated change?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Left some thoughts; also; could you update the commit message to describe the changes and new option in daemon.json?

Wondering; I see this option is currently not "live reloadable", which means that it's only possible to enable or disable buildkit by restarting the daemon. Should we make this option live-reloadable?

File: api/server/router/system/system_routes.go [hunk 17]
```diff
@@ -25,7 +25,10 @@ func optionsHandler(ctx context.Context, w http.ResponseWriter, r *http.Request,
 	return nil
 }
 
-func pingHandler(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+func (s *systemRouter) pingHandler(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+	if s.featureBuildKit != nil {
+		w.Header().Set("Feature-BuildKit-Enabled", fmt.Sprintf("%v", *s.featureBuildKit))
```
[the remark this answers is not in the document]
Comment (thaJeztah) [reply api/server/router/system/system_routes.go]: Oh! I just see it originally _was_ middleware. Was the main concern there that the header was added to every response @tiborvass? Wondering if we can make adding the header more isolated 🤔

**Comment by tiborvass** [comment]
> Should we make this option live-reloadable?

Yes.

File: api/server/router/system/system_routes.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (tiborvass) [reply api/server/router/system/system_routes.go]: Honestly I think it's fine to keep it in the ping handler. Middlewares are useful when they have to span across multiple handlers.

**Comment by thaJeztah** [comment]
>  if the daemon has buildkit enabled, it sets `Builder-Default-Version` header in the ping response, suggesting that it can be overridden. Otherwise, the header is not sent, and if buildkit was disabled but a client tries using buildkit on `/build` it will receive an error from the daemon.


Checking headers that are currently in `/_ping`;

```< HTTP/1.1 200 OK
< Api-Version: 1.39
< Docker-Experimental: false
< Ostype: linux
< Server: Docker/dev (linux)
< Date: Fri, 17 Aug 2018 16:32:03 GMT
< Content-Length: 2
< Content-Type: text/plain; charset=utf-8```

Wondering if we should omit `-Default` in `Builder-Default-Version`, so just `Builder-Version`. With buildkit _disabled_, it could return `Builder-Version: 1`, and no header should be considered equivalent to that?

File: api/types/types.go (hunk 15 above)
Comment (thaJeztah) [reply api/types/types.go]: Yes; Tibor explained that, thanks!

**Comment by AntaresS** [comment]
@tiborvass @vdemeester @cpuguy83 @thaJeztah lgty?

File: client/ping.go [hunk 18]
```diff
@@ -27,6 +27,9 @@ func (cli *Client) Ping(ctx context.Context) (types.Ping, error) {
 			ping.Experimental = true
 		}
 		ping.OSType = serverResp.header.Get("OSType")
+		if serverResp.header.Get("Builder-Version") == types.BuilderBuildKit {
```
Comment (tiborvass) [inline client/ping.go]: @AntaresS this should be:
```
if bv := serverResp.header.Get("Builder-Version"); bv != "" {
    ping.BuilderVersion = bv
}
```

File: client/ping.go [hunk 19]
```diff
@@ -7,7 +7,7 @@ import (
 	"github.com/docker/docker/api/types"
 )
 
-// Ping pings the server and returns the value of the "Docker-Experimental", "OS-Type" & "API-Version" headers
+// Ping pings the server and returns the value of the "Docker-Experimental", "Feature-BuildKit-Enabled", "OS-Type" & "API-Version" headers
```
Comment (tiborvass) [inline client/ping.go]: wrong header in comment

**Review by tiborvass — APPROVED** [review]
LGTM

**Comment by tiborvass** [comment]
@vdemeester @thaJeztah @cpuguy83 PTAL

File: cmd/dockerd/daemon.go [hunk 20]
```diff
@@ -293,16 +294,24 @@ func newRouterOptions(config *config.Config, daemon *daemon.Daemon) (routerOptio
...
 	if err != nil {
 		return opts, errors.Wrap(err, "failed to create buildmanager")
 	}
-
+	var bv types.BuilderVersion
+	if v, ok := config.Features["buildkit"]; ok {
+		if v {
+			bv = types.BuilderBuildKit
+		} else {
+			bv = types.BuilderV1
```
Comment (thaJeztah) [inline cmd/dockerd/daemon.go]: If the `buildkit` feature is not set in `daemon.json`, no `Builder-Version` header is set? I'd expect `features: {buildkit:false}}` and `{}` to be equivalent (i.e. in both cases buildkit is not active, so builder version is 1)

File: daemon/config/config.go [hunk 21]
```diff
@@ -56,6 +56,13 @@ var flatOptions = map[string]bool{
 	"default-ulimits":    true,
 }
 
+// skipValidateOptions contains configuration keys
+// that will be skipped from findConfigurationConflicts
+// for unknown flag validation.
+var skipValidateOptions = map[string]bool{
+	"feature-buildkit": true,
```
Comment (thaJeztah) [inline daemon/config/config.go]: This should be `buildkit` now?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
looks like there's still some issues with the current implementation/code:

```bash
$ mkdir -p /etc/docker
$ echo '{"experimental":true, "features": {"buildkit":true}}' > /etc/docker/daemon.json
$ dockerd
unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives don't match any configuration option: buildkit
```

Changing to `feature-buildkit` makes the daemon start;

```
$ echo '{"experimental":true, "features": {"feature-buildkit":true}}' > /etc/docker/daemon.json
...
INFO[2018-08-19T09:12:24.248409702Z] Daemon has completed initialization
INFO[2018-08-19T09:12:24.274417246Z] API listen on /var/run/docker.sock
```

But buildkit won't be enabled / visible in `/_ping`;

```bash
$ curl -v --unix-socket /var/run/docker.sock http://localhost/_ping
...
< HTTP/1.1 200 OK
< Api-Version: 1.39
< Docker-Experimental: true
< Ostype: linux
< Server: Docker/dev (linux)
< Date: Sun, 19 Aug 2018 09:13:17 GMT
< Content-Length: 2
< Content-Type: text/plain; charset=utf-8
```

Live-reload also isn't implemented yet (could possibly be done in a follow-up);

```bash
$ rm /etc/docker/daemon.json 
$ kill -HUP $(pidof dockerd)
```

```
Got signal to reload configuration, reloading from: /etc/docker/daemon.json 
INFO[2018-08-19T09:18:19.584233372Z] Reloaded configuration: {"mtu":1500,"pidfile":"/var/run/docker.pid","data-root":"/var/lib/docker","exec-root":"/var/run/docker","group":"docker","deprecated-key-path":"/etc/docker/key.json","max-concurrent-downloads":3,"max-concurrent-uploads":5,"shutdown-timeout":15,"hosts":["unix:///var/run/docker.sock"],"log-level":"info","swarm-default-advertise-addr":"","swarm-raft-heartbeat-tick":0,"swarm-raft-election-tick":0,"metrics-addr":"","log-driver":"json-file","ip":"0.0.0.0","icc":true,"iptables":true,"ip-forward":true,"ip-masq":true,"userland-proxy":true,"default-address-pools":{},"network-control-plane-mtu":1500,"disable-legacy-registry":true,"experimental":true,"containerd":"/var/run/docker/containerd/docker-containerd.sock","features":{"feature-buildkit":true},"runtimes":{"runc":{"path":"docker-runc"}},"default-runtime":"runc","oom-score-adjust":-500,"default-shm-size":67108864,"default-ipc-mode":"shareable","resolv-conf":"/etc/resolv.conf"} 
```

File: client/ping.go [hunk 22]
```diff
@@ -27,6 +27,9 @@ func (cli *Client) Ping(ctx context.Context) (types.Ping, error) {
 			ping.Experimental = true
 		}
 		ping.OSType = serverResp.header.Get("OSType")
+		if bv := serverResp.header.Get("Builder-Version"); bv != "" {
```
Comment (thaJeztah) [inline client/ping.go]: This should also be documented in the swagger; https://github.com/moby/moby/blob/896d1b1c61a48e2df1a7b4644ddde6ee97db6111/api/swagger.yaml?utf8=✓#L6967-L6985

File: daemon/config/config.go (hunk 21 above)
Comment (tiborvass) [reply daemon/config/config.go]: Yes, I thought this was fixed. @AntaresS

File: cmd/dockerd/daemon.go (hunk 20 above)
Comment (tiborvass) [reply cmd/dockerd/daemon.go]: @thaJeztah yes. So it doesn't matter.

File: daemon/config/config.go (hunk 21 above)
Comment (tiborvass) [reply daemon/config/config.go]: Or maybe just `features`

Comment (AntaresS) [reply daemon/config/config.go]: done

File: client/ping.go (hunk 22 above)
Comment (AntaresS) [reply client/ping.go]: done

File: cmd/dockerd/daemon.go (hunk 20 above)
Comment (AntaresS) [reply cmd/dockerd/daemon.go]: @thaJeztah We actually meant to make it distinguished, so `features: {buildkit:false}}` means `buildkit` is not allowed while `{}` (or not set in daemon.json)  means it's up to client to choose builder version. For header it is the same concept.

Comment (thaJeztah) [reply cmd/dockerd/daemon.go]: Still confused by this; how would the client know that buildkit is available?

Comment (AntaresS) [reply cmd/dockerd/daemon.go]: @thaJeztah It's obvious that when the `features` is set in `daemon.json` client will know the version. If it is not set in the config file (i.e. header will not include `Builder-Version` either), `cli` will use the env var to determine whether to make buildkit request or otherwise use the old builder. Maybe the changes from cli can also help you - https://github.com/docker/cli/pull/1275

Comment (thaJeztah) [reply cmd/dockerd/daemon.go]: Got it

**Review by thaJeztah — APPROVED** [review]
LGTM
